### DISCUSSION TITLE
Clarifications on usage of CallbackParam

### DISCUSSION THREAD

**Discussion by marchinidavide** [cvxpy/cvxpy#2458] [Q&A]
Hello, 
I'm trying to understand if I can make more use of `CallbackParam` but I'm not sure I get their usage in a correct way.
It seems to just work magically (so huge kudos!) but I don't really undestand why.
Specifically, how is the code able to parametrize the solving chain through the CallbackParam? Any material I can dig more into?
A self contained small example below, solving linear regression when the features matrix is the product of two matrices.

This says my problem has 4 variables (expected) and 50 parameters (which is `10x4` shape of `P0` + `10` size of `b`) while I would have expected the number of parameters to be 94: `10x6` shape of `A1` + `6x4` shape of `A2` + `10` size of `b`. What am I missing?

```python
import cvxpy as cvx
import numpy as np

rng = np.random.default_rng(seed=0)
x = cvx.Variable(4)
b = cvx.Parameter(10, value=rng.standard_normal(10))
A1 = cvx.Parameter((10, 6), value=rng.standard_normal((10, 6)))
A2 = cvx.Parameter((6, 4), value=rng.standard_normal((6, 4)))
P0 = cvx.CallbackParam(
    callback=lambda: A1.value @ A2.value,
    shape=(10, 4),
)
p = cvx.Problem(objective=cvx.Minimize(cvx.sum_squares(b - P0 @ x)))
v = p.solve(
    enforce_dpp=True,
    solver="CLARABEL",
    verbose=True,
    solver_verbose=False,
)
assert p.status == "optimal"
sol0 = np.linalg.lstsq(a=A1.value @ A2.value, b=b.value, rcond=-1)
print(x.value, v)
print(sol0[0], sol0[1][0])

A1.value += 1.0
v = p.solve(
    enforce_dpp=True,
    solver="CLARABEL",
    verbose=True,
    solver_verbose=False,
)
assert p.status == "optimal"
sol0 = np.linalg.lstsq(a=A1.value @ A2.value, b=b.value, rcond=-1)
print(x.value, v)
print(sol0[0], sol0[1][0])
```

More broadly, is there any drawback of using `CallbackParam` (in terms of solver speed/accuracy)?

**Comment by SteveDiamond** [comment]
The CallbackParam works by evaluating the callback to get the parameter value right before the solver is called. So cvxpy doesn't know anything about A1 and A2 since they are only referenced in the callback. cvxpy only knows about parameters that are in the explicit problem formulation/expression trees.

**Reply by marchinidavide** [reply]
Thank you Steve!
Does this mean that the matmul inside the callback (`A1.value @ A2.value`) is executed each time i call `.solve` or only when I change the value of either `A1` or `A2`?

**Reply by marchinidavide** [reply]
Answering my own question: in this example the callback is called 4 times each time the problem is solved, no matter if i change something in any of the parameter or no change at all. I guess in a big problem this can hurt performance.
